### PR TITLE
Enable valgrind tests on AArch64 by setting ARMCAP via static defines to avoid reading from MIDR_EL1.

### PR DESCRIPTION
**Pull request by nebeid**
### Issues:
V886376127

### Description of changes: 
Enable Valgrind tests on Arm as a follow-up to https://github.com/aws/aws-lc/pull/970.
Valgrind was choking (illegal opcode) on https://github.com/aws/aws-lc/blob/300b5eef6753824b502d7c0e35dcd38567e82e7c/crypto/fipsmodule/cpucap/cpu_aarch64_linux.c#L39

Despite that this issue is [said to be resolved](https://bugs.kde.org/show_bug.cgi?id=381556), versions 3.19 and 3.20 were outputting the same error of "illegal opcode". (Note: The newer versions solved the issue we had with reading the capabilities at runtime via `getauxval()`. But then we added to that getting the CPU part as mentioned above.)

Our workaround was to avoid compiling `cpu_aarch64_linux.c` by passing `-DOPENSSL_STATIC_ARMCAP` to the test build and setting all capabilities via the STATIC_ARMCAP macros (except SHA512 which wasn't used before with valgrind and was also causing an "illegal opcode" error when set).
Note: Setting `-DOPENSSL_STATIC_ARMCAP` without specifying any capabilities disables all of them as in the second 
set of tests.

Also, the ssl tests were made run in parallel to the valgrind tests on libcrypto on a separate machine.

### Call-outs:
* Timeout for ssl tests was increased to 60 min (from 10 min), due to valgrind requiring more time.
* ssl tests are performed only with all capabilities enabled; and not when they are disabled.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and 
the ISC license.

### CODE REVIEW

**Comment by nebeid** [comment]
I disabled SHA512 after I tried running tests/ci/run_valgrind_tests.sh on [v 1.8.0](https://github.com/aws/aws-lc/releases/tag/v1.8.0) with the following line
https://github.com/aws/aws-lc/blame/b78d997395d543e557b1309dd9f4b3522b3a4118/util/all_tests.json#L27
changed to
```
"env": ["OPENSSL_armcap=0x7D"]
```
and getting the following response
```
crypto/crypto_test [shard 4/4]
crypto/crypto_test failed to complete: exit status 99
crypto/crypto_test [shard 3/4]
crypto/crypto_test [shard 1/4] (custom environment)
crypto/crypto_test [shard 2/4] (custom environment)
crypto/crypto_test [shard 3/4] (custom environment)
crypto/crypto_test [shard 4/4] (custom environment)
Fatal Error: HW capability found: 0x3D, but HW capability requested: 0x7D.
OPENSSL_armcap=0x7D crypto/crypto_test [shard 1/4]
crypto/crypto_test failed to complete: exit status 1
Fatal Error: HW capability found: 0x3D, but HW capability requested: 0x7D.
OPENSSL_armcap=0x7D crypto/crypto_test [shard 2/4]
crypto/crypto_test failed to complete: exit status 1
Fatal Error: HW capability found: 0x3D, but HW capability requested: 0x7D.
OPENSSL_armcap=0x7D crypto/crypto_test [shard 3/4]
crypto/crypto_test failed to complete: exit status 1
Fatal Error: HW capability found: 0x3D, but HW capability requested: 0x7D.
OPENSSL_armcap=0x7D crypto/crypto_test [shard 4/4]
```

**Comment by nebeid** [comment]
###Test
Test that XTS buffer overread https://github.com/aws/aws-lc/commit/7c3c62c87b1b8f8d7668bf5856651d1927651f84 is caught
* Bring it back
```
diff --git a/crypto/fipsmodule/aes/asm/aesv8-armx.pl b/crypto/fipsmodule/aes/asm/aesv8-armx.pl
index 1bc05826f..836cd28ec 100644
--- a/crypto/fipsmodule/aes/asm/aesv8-armx.pl
+++ b/crypto/fipsmodule/aes/asm/aesv8-armx.pl
@@ -2109,7 +2109,7 @@ $code.=<<___    if ($flavour =~ /64/);
 .align  4
 .Lxts_dec_tail4x:
     add     $inp,$inp,#16
-    tst     $tailcnt,#0xf
+    vld1.32 {$dat0},[$inp],#16
     veor    $tmp1,$dat1,$tmp0
     vst1.8  {$tmp1},[$out],#16
     veor    $tmp2,$dat2,$tmp2
@@ -2118,8 +2118,6 @@ $code.=<<___    if ($flavour =~ /64/);
     veor    $tmp4,$dat4,$tmp4
     vst1.8  {$tmp3-$tmp4},[$out],#32
 
-    b.eq    .Lxts_dec_abort
-    vld1.32 {$dat0},[$inp],#16
     b       .Lxts_done
 .align  4
 .Lxts_outer_dec_tail:
```
* Run in the same docker image as CI
```
bash-5.2# cmake '-DCMAKE_C_FLAGS=-DOPENSSL_STATIC_ARMCAP -DOPENSSL_STATIC_ARMCAP_NEON -DOPENSSL_STATIC_ARMCAP_AES -DOPENSSL_STATIC_ARMCAP_PMULL  -DOPENSSL_STATIC_ARMCAP_SHA1 -DOPENSSL_STATIC_ARMCAP_SHA256  -DOPENSSL_STATIC_ARMCAP_SHA3 -DOPENSSL_STATIC_ARMCAP_NEOVERSE_V1' -GNinja ..

bash-5.2# valgrind --error-exitcode=99 --track-origins=yes --leak-check=full --trace-children=yes crypto/crypto_test "--gtest_filter=XTSTest.*" 
==4539== Memcheck, a memory error detector
==4539== Copyright (C) 2002-2022, and GNU GPL'd, by Julian Seward et al.
==4539== Using Valgrind-3.19.0 and LibVEX; rerun with -h for copyright info
==4539== Command: crypto/crypto_test --gtest_filter=XTSTest.*
==4539== 
Note: Google Test filter = XTSTest.*
[==========] Running 4 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 4 tests from XTSTest
[ RUN      ] XTSTest.TestVectors
==4539== Invalid read of size 16
==4539==    at 0x8CFD14: aes_hw_xts_decrypt (aesv8-armx.S:1875)
==4539==  Address 0x538a3f0 is 0 bytes after a block of size 144 alloc'd
==4539==    at 0x48A3FCC: operator new(unsigned long) (vg_replace_malloc.c:422)
==4539==    by 0x4EC3D3: __gnu_cxx::new_allocator<unsigned char>::allocate(unsigned long, void const*) (new_allocator.h:127)
==4539==    by 0x4EA7D3: std::allocator_traits<std::allocator<unsigned char> >::allocate(std::allocator<unsigned char>&, unsigned long) (alloc_traits.h:464)
==4539==    by 0x4E7F47: std::_Vector_base<unsigned char, std::allocator<unsigned char> >::_M_allocate(unsigned long) (stl_vector.h:346)
==4539==    by 0x7D05F7: std::vector<unsigned char, std::allocator<unsigned char> >::reserve(unsigned long) (vector.tcc:78)
==4539==    by 0x7D0397: DecodeHex(std::vector<unsigned char, std::allocator<unsigned char> >*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (test_util.cc:47)
==4539==    by 0x68EB33: XTSTest_TestVectors_Test::TestBody() (xts_test.cc:630)
==4539==    by 0x94405F: void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) (gtest.cc:2605)
==4539==    by 0x93E99F: void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) (gtest.cc:2660)
==4539==    by 0x9230AF: testing::Test::Run() (gtest.cc:2680)
==4539==    by 0x9237A7: testing::TestInfo::Run() (gtest.cc:2858)
==4539==    by 0x923E8B: testing::TestSuite::Run() (gtest.cc:3012)
==4539== 
[       OK ] XTSTest.TestVectors (110 ms)
[ RUN      ] XTSTest.DuplicateKey
[       OK ] XTSTest.DuplicateKey (3 ms)
[ RUN      ] XTSTest.InputTooLong
[       OK ] XTSTest.InputTooLong (6 ms)
[ RUN      ] XTSTest.SectorTweakAPIUsage
[       OK ] XTSTest.SectorTweakAPIUsage (24 ms)
[----------] 4 tests from XTSTest (151 ms total)

[----------] Global test environment tear-down
[==========] 4 tests from 1 test suite ran. (186 ms total)
[  PASSED  ] 4 tests.
==4539== 
==4539== HEAP SUMMARY:
==4539==     in use at exit: 440 bytes in 2 blocks
==4539==   total heap usage: 34,156 allocs, 34,154 frees, 3,609,309 bytes allocated
==4539== 
==4539== LEAK SUMMARY:
==4539==    definitely lost: 0 bytes in 0 blocks
==4539==    indirectly lost: 0 bytes in 0 blocks
==4539==      possibly lost: 0 bytes in 0 blocks
==4539==    still reachable: 440 bytes in 2 blocks
==4539==         suppressed: 0 bytes in 0 blocks
==4539== Reachable blocks (those to which a pointer was found) are not shown.
==4539== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==4539== 
==4539== For lists of detected and suppressed errors, rerun with: -s
==4539== ERROR SUMMARY: 4 errors from 1 contexts (suppressed: 0 from 0)
bash-5.2# 
```

File: tests/ci/common_posix_setup.sh [hunk 1]
```diff
@@ -140,8 +156,31 @@ function fips_build_and_test {
...
+    #  in cpu_aarch64_linux.c)
+    run_build "$@" -DCMAKE_C_FLAGS="-DOPENSSL_STATIC_ARMCAP"
+    run_cmake_custom_target 'run_tests_valgrind'
+  else
+    run_build "$@"
+    run_cmake_custom_target 'run_tests_valgrind'
+  fi
+}
+
+function build_and_test_ssl_valgrind {
```
Comment (samuel40791765) [inline tests/ci/common_posix_setup.sh:175]: This only seems to be running the ssl runner tests in valgrind , while `build_and_test_valgrind` still calls both `crypto_test` and `ssl_test`.
With that in context, could we rename this into something clearer? I got a bit confused and thought this was running `ssl_test` as well. `build_and_test_ssl_runner_valgrind` might be good.

File: tests/ci/cdk/cdk/codebuild/github_ci_linux_arm_omnibus.yaml [hunk 2]
```diff
@@ -289,10 +289,18 @@ batch:
...
+    - identifier: amazonlinux2023_gcc11x_aarch_valgrind
+      buildspec: ./tests/ci/codebuild/common/run_valgrind_tests.yml
+      env:
+        type: ARM_CONTAINER
+        privileged-mode: true
+        compute-type: BUILD_GENERAL1_LARGE
+        image: 620771051181.dkr.ecr.us-west-2.amazonaws.com/aws-lc-docker-images-linux-aarch:amazonlinux-2023_gcc-11x_latest
+
+    - identifier: amazonlinux2023_gcc11x_aarch_ssl_valgrind
+      buildspec: ./tests/ci/codebuild/common/run_ssl_valgrind_tests.yml
```
Comment (samuel40791765) [inline tests/ci/cdk/cdk/codebuild/github_ci_linux_arm_omnibus.yaml:301]:
```suggestion
    - identifier: amazonlinux2023_gcc11x_aarch_ssl_runner_valgrind
      buildspec: ./tests/ci/codebuild/common/run_ssl_runner_valgrind_tests.yml
```
Same comments here about clarifying this is just for the ssl runner tests.

Comment (nebeid) [reply tests/ci/cdk/cdk/codebuild/github_ci_linux_arm_omnibus.yaml]: Done.

File: tests/ci/common_posix_setup.sh (hunk 1 above)
Comment (nebeid) [reply tests/ci/common_posix_setup.sh]: Thanks, Samuel, for the clarification. Done.

**Review by samuel40791765 — APPROVED** [review]

**Review by skmcgrail — APPROVED** [review]

**Comment by nebeid** [comment]
Related to this PR: PR #1006 Check HWCAP_CPUID before reading MIDR_EL1 on aarch64. It seems that Valgrind's [fix to the issue](https://bugs.kde.org/show_bug.cgi?id=381556), was to mask HWCAP_CPUID as was suggested in the [workaround here](https://bugs.launchpad.net/ubuntu/+source/valgrind/+bug/1826811).
That PR would have fixed the issue we had here with Valgrind but wouldn't enable the code that runs on CPUs for which we check the parts (Neoverse V1, so far). To enable this code, this PR sets the CPU part at compile time bypassing runtime checks.

**Pull request merged**